pkg/gatherers/conditional/conditional_gatherer.go

```diff
@@ -0,0 +1,380 @@
+// they can be fetched from outside, checked that they make sense (we want to check the parameters, for example if
+// a rule tells to collect logs of a namespace on firing alert, we want to check that the namespace is created
+// by openshift and not by a user). Conditional gathering isn't considered prioritized, so we run it every 6 hours.
+package conditional
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/insights-operator/pkg/gatherers"
+	"github.com/openshift/insights-operator/pkg/gatherers/common"
+	"github.com/openshift/insights-operator/pkg/record"
+	"github.com/openshift/insights-operator/pkg/utils"
+)
+
+// gatheringFunctionBuilders lists all the gatherers which can be run on some condition. Gatherers can have parameters,
+// like namespace or number of log lines to fetch, see the docs of the functions.
+var gatheringFunctionBuilders = map[gatheringFunctionName]gathererFunctionBuilderPtr{
+	gatherLogsOfNamespace:         (*Gatherer).BuildGatherLogsOfNamespace,
+	gatherImageStreamsOfNamespace: (*Gatherer).BuildGatherImageStreamsOfNamespace,
+}
+
+// gatheringRules contains all the rules used to run conditional gatherings.
+// Right now they are declared here in the code, but later they can be fetched from an external source.
+// An example of the future json version of this is:
+//   [{
+//     "conditions": [
+//       {
+//         "type": "alert_is_firing",
+//         "params": {
+//           "name": "ClusterVersionOperatorIsDown"
+//         }
+//       },
+//       {
+//         "type": "cluster_version_equals",
+//         "params": {
+//           "version": "4.8"
+//         }
+//       }
+//     ],
+//     "gathering_functions": {
+//       "gather_logs_of_namespace": {
+//         "namespace": "openshift-cluster-version",
+//         "keep_lines": 100
+//       },
+//     }
+//   }]
+// Which means to collect logs of all containers from all pods in namespace openshift-monitoring keeping last 100 lines
+// per container only if cluster version is 4.8 (not implemented, just an example of possible use) and alert
+// ClusterVersionOperatorIsDown is firing
+var gatheringRules = []gatheringRule{
+	{
+		Conditions: []conditionWithParams{
+			{
+				Type: alertIsFiring,
+				Params: map[string]interface{}{
+					"name": "SamplesImagestreamImportFailing",
+				},
+			},
+		},
+		GatheringFunctions: map[gatheringFunctionName]GatheringFunctionParams{
+			gatherLogsOfNamespace: map[string]interface{}{
+				"namespace":      "openshift-cluster-samples-operator",
+				"tail_lines":     100,
+				"label_selector": "",
+			},
+			gatherImageStreamsOfNamespace: map[string]interface{}{
+				"namespace": "openshift-cluster-samples-operator",
+			},
+		},
+	},
+}
+
+const (
+	canConditionalGathererFail = false
+	conditionalGathererPeriod  = time.Hour * 6
```

Comment:
Why 6 hours? I think it would make sense to use the default/configured interval (i.e the same as in `clusterconfig` gatherer)

Reply:
Well, later we may load the gathering rules from an external API, so there may be many things to gather. But I think it's a question to the PMs, we can as well collect it as often as `clusterconfig` gatherer, it all depends on how fast we want to get this conditionally gathered data.

Reply:
My thinking is following. Alerts are gathered in `clusterconfig` anyway so from data analysis point of view you will see that an alert XY is firing, but you will not have related conditional data, because the `conditional` gatherer didn't run this time. This seems to me confusing.

Reply:
Yeah, but for the conditional gatherer we can look at the metadata if it actually collected something for this archive. No problem removing it, tho, it doesn't make much difference, just the conditional gatherer will run more often.